designs/data_serialization_design.md

```diff
@@ -0,0 +1,189 @@
+Chosen option: Protocol Buffers (protobuf), due to:
+- Binary format
+- Native support in gRPC
+- Strong typing and schema evolution
+- Built-in JSON support (for debugging)
+- Schema Design
+
+The schema has two broad parts:
+1)  Config (TrackerCfg) — defined by customers, drives tracker creation.
+2) Runtime Data — driven by traffic and execution:
+   - Consts: Initialized once, updated periodically (e.g., hash seed).
+   - State: Updated frequently on the hot path (bucket data).
+
+Accessing `State` requires the matching `Config` and `Consts`. Multiple tracker configs may exist simultaneously (multi-tenant).
+
+## Schema v1 (High-level)
+```
+package fair.data.v1;
+
+enum LevelSquashingFunction {
+  LEVEL_SQUASHING_FUNCTION_UNSPECIFIED = 0;
+  LEVEL_SQUASHING_FUNCTION_MIN = 1;
+  LEVEL_SQUASHING_FUNCTION_MEAN = 2;
+}
+
+enum HostSquashingFunction {
+  HOST_SQUASHING_FUNCTION_UNSPECIFIED = 0;
+  HOST_SQUASHING_FUNCTION_LOGIC1 = 1;
+  HOST_SQUASHING_FUNCTION_LOGIC2 = 2;
+}
+
+message TrackerCfg {
+  int64 config_guid = 1; // Unique per tracker config
+  int64 config_version = 2; // Monotonically increments per version of tracker config
+  uint32 m = 3;
+  uint32 l = 4;
+  double pi = 5;
+  double pd = 6;
+  double lambda = 7;
+  google.protobuf.Duration rotation_frequency = 8;
+  LevelSquashingFunction level_squash_fn = 9;
+  HostSquashingFunction host_squash_fn = 10;
+}
+
+message Bucket {
+  double probability = 1;
+  uint64 last_updated_time_ms = 2;
+}
+
+message Level {
+  repeated Bucket buckets = 1;
+}
+
+message FairData {
+  repeated Level levels = 1;
+}
+
+message FairRunConsts {
+  uint32 murmur_seed = 1;
+  string ref_guid = 2;
+}
+
+message HostMeta {
+  string host_guid = 1;
+  uint64 serialized_at_ms = 2;
+}
+
+message FairRunTimeData {
+  FairRunConsts runtime = 1;
+  FairData data = 2;
+}
+
+message FairSt {
+  // int32 version_num = 1; May be do this only if necessary
+  repeated TrackerCfg cfgs = 2;
+  repeated FairRunTimeData data = 3;
+  HostMeta meta = 4;
+}
+```
+
+The top-level struct contains a list of `TrackerCfgs`, the Runtime data is encoded using `FairRunTimeData` - which contains the consts and the data. Each unit of the run-time data contains a link `ref_guid` which refers to the config_guid and version. The structure also include a `HostMeta` field that includes information about the host.
+
+
+## Implementation Plan
+### Phase 1: Core
+
+Define schema (proto/fair/data/v1/structure.proto).
+Generate Go code with protoc-gen-go.
+Implement:
+```
+func (s *Structure) Serialize() ([]byte, error)
+func DeserializeStructure(data []byte) (*Structure, error)
+```
+
+### Phase 2: Integration
+- Add serialization tests
+- Validate deserialized data
+- Robust error handling
+- Benchmark performance 
+- Support human readable dumping (JSON)
+
+### Phase 3: Enhancements
+- Optional compression (gzip/lz4).
```

Comment:
If we are doing it, I'll vote `zstd` but again, we are a constant space library so I really don't think that's a problem.